File: src/main/python/preponderous/viron/services/environmentService.py

# Copyright (c) 2024 Preponderous Software
# MIT License

from typing import Optional
import requests
from src.main.python.preponderous.viron.models.environment import Environment

class EnvironmentService:
    def __init__(self, viron_host: str, viron_port: int, auth_token: Optional[str] = None):
        self.viron_host = viron_host
        self.viron_port = viron_port
        self.auth_token = auth_token

    def get_base_url(self) -> str:
        return f"{self.viron_host}:{self.viron_port}/api/v1/environments"

    def get_auth_headers(self) -> dict:
        return {"Authorization": f"Bearer {self.auth_token}"} if self.auth_token else {}

    def get_all_environments(self) -> list[Environment]:
        response = requests.get(f"{self.get_base_url()}", headers=self.get_auth_headers())
        response.raise_for_status()
        return [Environment(**env) for env in response.json()]

    def get_environment_by_id(self, environment_id: int) -> Environment:
        response = requests.get(f"{self.get_base_url()}/{environment_id}", headers=self.get_auth_headers())
        response.raise_for_status()
        return Environment(**response.json())

    def get_environment_by_name(self, name: str) -> Environment:
        response = requests.get(f"{self.get_base_url()}/name/{name}", headers=self.get_auth_headers())
        response.raise_for_status()
        return Environment(**response.json())

    def get_environment_of_entity(self, entity_id: int) -> Environment:
        response = requests.get(f"{self.get_base_url()}/entity/{entity_id}", headers=self.get_auth_headers())
        response.raise_for_status()
        return Environment(**response.json())

    def create_environment(self, name: str, num_grids: int, grid_size: Optional[int] = None,
                           num_rows: Optional[int] = None, num_columns: Optional[int] = None) -> Environment:
        """Create an environment containing num_grids grids.

        Supply either grid_size (square shorthand) or both num_rows and num_columns.
        num_rows/num_columns take precedence over grid_size; supplying only one of them
        is rejected by the API with a 400.
        """
        if (num_rows is None) != (num_columns is None):
            raise ValueError("num_rows and num_columns must be provided together")
        if num_rows is None and grid_size is None:
            raise ValueError("either grid_size or both num_rows and num_columns must be provided")

        payload = {"name": name, "numGrids": num_grids}
        if grid_size is not None:
            payload["gridSize"] = grid_size
        if num_rows is not None:
            payload["numRows"] = num_rows
        if num_columns is not None:
            payload["numColumns"] = num_columns

        response = requests.post(f"{self.get_base_url()}", json=payload, headers=self.get_auth_headers())
        response.raise_for_status()
        return Environment(**response.json())

    def delete_environment(self, environment_id: int) -> bool:
        response = requests.delete(f"{self.get_base_url()}/{environment_id}", headers=self.get_auth_headers())
        response.raise_for_status()
        return True

    def update_environment_name(self, environment_id: int, name: str) -> bool:
        response = requests.patch(f"{self.get_base_url()}/{environment_id}/name", json={"name": name}, headers=self.get_auth_headers())
        response.raise_for_status()
        return response.status_code == 200

File: src/main/java/preponderous/viron/services/EnvironmentService.java
// Copyright (c) 2024 Preponderous Software
// MIT License

package preponderous.viron.services;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

import lombok.extern.slf4j.Slf4j;
import preponderous.viron.config.AuthTokenInterceptor;
import preponderous.viron.config.ServiceConfig;
import preponderous.viron.dto.CreateEnvironmentRequest;
import preponderous.viron.dto.UpdateEnvironmentNameRequest;
import preponderous.viron.exceptions.NotFoundException;
import preponderous.viron.exceptions.ServiceException;
import preponderous.viron.models.Environment;

@Service
@Slf4j
public class EnvironmentService {
    private final RestTemplateBuilder restTemplateBuilder;
    private final ServiceConfig serviceConfig;
    private final String baseUrl;

    @Autowired
    public EnvironmentService(RestTemplateBuilder restTemplateBuilder, ServiceConfig serviceConfig) {
        this.restTemplateBuilder = restTemplateBuilder.additionalInterceptors(new AuthTokenInterceptor(serviceConfig));
        this.serviceConfig = serviceConfig;
        this.baseUrl = this.serviceConfig.getVironHost() + ":" + serviceConfig.getVironPort() + "/api/v1/environments";
    }

    public List<Environment> getAllEnvironments() {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Environment[]> response = restTemplate.exchange(baseUrl, HttpMethod.GET, null, Environment[].class);
            return response.getStatusCode() == HttpStatus.OK && response.getBody() != null
                    ? Arrays.asList(response.getBody())
                    : Collections.emptyList();
        } catch (Exception e) {
            log.error("Error getting environments: {}", e.getMessage());
            throw new ServiceException("Error getting environments", e);
        }
    }

    public Environment getEnvironmentById(int id) {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Environment> response = restTemplate.exchange(baseUrl + "/{id}", HttpMethod.GET, null, Environment.class, id);
            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                return response.getBody();
            }
            throw new NotFoundException("Environment not found with id: " + id);
        } catch (HttpClientErrorException.NotFound e) {
            throw new NotFoundException("Environment not found with id: " + id);
        } catch (ServiceException e) {
            throw e;
        } catch (Exception e) {
            log.error("Error getting environment by id {}: {}", id, e.getMessage());
            throw new ServiceException("Error getting environment by id: " + id, e);
        }
    }

    public Environment getEnvironmentByName(String name) {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Environment> response = restTemplate.exchange(baseUrl + "/name/{name}", HttpMethod.GET, null, Environment.class, name);
            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                return response.getBody();
            }
            throw new NotFoundException("Environment not found with name: " + name);
        } catch (HttpClientErrorException.NotFound e) {
            throw new NotFoundException("Environment not found with name: " + name);
        } catch (ServiceException e) {
            throw e;
        } catch (Exception e) {
            log.error("Error getting environment by name {}: {}", name, e.getMessage());
            throw new ServiceException("Error getting environment by name: " + name, e);
        }
    }

    public Environment getEnvironmentOfEntity(int entityId) {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Environment> response = restTemplate.exchange(baseUrl + "/entity/{entityId}", HttpMethod.GET, null, Environment.class, entityId);
            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                return response.getBody();
            }
            throw new NotFoundException("Environment not found for entity: " + entityId);
        } catch (HttpClientErrorException.NotFound e) {
            throw new NotFoundException("Environment not found for entity: " + entityId);
        } catch (ServiceException e) {
            throw e;
        } catch (Exception e) {
            log.error("Error getting environment of entity {}: {}", entityId, e.getMessage());
            throw new ServiceException("Error getting environment of entity: " + entityId, e);
        }
    }

    /**
     * Creates an environment whose grids are square.
     *
     * @param gridSize rows and columns of each grid
     */
    public Environment createEnvironment(String name, int numGrids, int gridSize) {
        return createEnvironment(new CreateEnvironmentRequest(name, numGrids, gridSize));
    }

    /**
     * Creates an environment whose grids have independently sized dimensions.
     *
     * @param numRows    rows in each grid
     * @param numColumns columns in each grid
     */
    public Environment createEnvironment(String name, int numGrids, int numRows, int numColumns) {
        return createEnvironment(new CreateEnvironmentRequest(name, numGrids, null, numRows, numColumns));
    }

    private Environment createEnvironment(CreateEnvironmentRequest request) {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Environment> response = restTemplate.exchange(
                    baseUrl,
                    HttpMethod.POST,
                    new HttpEntity<>(request),
                    Environment.class
            );
            if (response.getStatusCode().is2xxSuccessful() && response.getBody() != null) {
                return response.getBody();
            }
            throw new ServiceException("Error creating environment");
        } catch (ServiceException e) {
            throw e;
        } catch (Exception e) {
            log.error("Error creating environment {}: {}", request.getName(), e.getMessage());
            throw new ServiceException("Error creating environment", e);
        }
    }

    public boolean deleteEnvironment(int id) {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Void> response = restTemplate.exchange(baseUrl + "/{id}", HttpMethod.DELETE, null, Void.class, id);
            if (response.getStatusCode() == HttpStatus.NOT_FOUND) {
                throw new NotFoundException("Environment not found with id: " + id);
            }
            if (!response.getStatusCode().is2xxSuccessful()) {
                throw new ServiceException("Error deleting environment");
            }
            return true;
        } catch (HttpClientErrorException.NotFound e) {
            throw new NotFoundException("Environment not found with id: " + id);
        } catch (ServiceException e) {
            throw e;
        } catch (Exception e) {
            log.error("Error deleting environment {}: {}", id, e.getMessage());
            throw new ServiceException("Error deleting environment", e);
        }
    }

    public boolean updateEnvironmentName(int id, String name) {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Void> response = restTemplate.exchange(
                    baseUrl + "/{id}/name",
                    HttpMethod.PATCH,
                    new HttpEntity<>(new UpdateEnvironmentNameRequest(name)),
                    Void.class,
                    id
            );
            if (response.getStatusCode() == HttpStatus.NOT_FOUND) {
                throw new NotFoundException("Environment not found with id: " + id);
            }
            if (!response.getStatusCode().is2xxSuccessful()) {
                throw new ServiceException("Error updating environment name");
            }
            return true;
        } catch (HttpClientErrorException.NotFound e) {
            throw new NotFoundException("Environment not found with id: " + id);
        } catch (ServiceException e) {
            throw e;
        } catch (Exception e) {
            log.error("Error updating name for environment {}: {}", id, e.getMessage());
            throw new ServiceException("Error updating environment name", e);
        }
    }
}
